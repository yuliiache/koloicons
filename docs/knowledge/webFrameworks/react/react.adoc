== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







