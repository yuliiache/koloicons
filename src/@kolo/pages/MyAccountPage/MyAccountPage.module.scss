@import "@kolo/design";

.header {
  padding: 0.8rem 0;
  background-color: $WhiteBG;

  &Container {
    margin: 0 auto;
    padding: 0 0.8rem;
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    max-width: 34.1rem;

    @include mobile-screen {
      flex-wrap: nowrap;
    }
  }
}

.circle {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 2.4rem;
  height: 2.4rem;
  border-radius: 50%;
  border: 0.2rem solid $BlueColor;
  background-color: $LightBrandColor;
  color: $BlueColor;
  font-size: 0.9rem;
  margin-right: 0.4rem;
}

.userInfo {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  margin-right: auto;
  margin-left: 0.8rem;

  h1 {
    color: $secFontColor;
    font-weight: 500;
    font-size: 0.8rem;
  }

  p {
    width: 100%;
    font-size: 0.6rem;
    line-height: 1.4;
    color: $GreyColor;
  }

  @include mobile-screen {
    h1 {
      font-size: 1.2rem;
    }
  }
}

.inputGroup {
  max-width: 34.1rem;
  padding: 2.2rem 0.8rem 2.4rem;
  margin: auto;
}
