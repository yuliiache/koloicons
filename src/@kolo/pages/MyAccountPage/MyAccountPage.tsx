import Input from '@kolo/uiKit/Input/Input';
import {InputType} from '@kolo/uiKit/Input/InputType';
import React from 'react';

import styles from './MyAccountPage.module.scss';

interface User {
  firstName: string;
  lastName: string;
  email: string;
}

const MyAccountPage: React.FC = () => {
  const testUser: User = {
    firstName: 'John',
    lastName: 'Doe',
    email: 'john.doe@mail.com',
  };

  const getFirstLetter = (str: string): string => {
    return str.charAt(0);
  };

  const fullName = `${testUser.firstName} ${testUser.lastName}`;

  const handleChange = (e: string): void => {
    console.log(e);
  };

  return (
    <main>
      <section className={styles.header}>
        <div className={styles.headerContainer}>
          <div className={styles.circle}>{getFirstLetter(testUser.firstName)}</div>
          <section className={styles.userInfo}>
            <h1>{testUser.firstName}</h1>
            <p>{testUser.email}</p>
          </section>
        </div>
      </section>
      <div className={styles.inputGroup}>
        <Input
          id={'name'}
          value={fullName}
          onChange={handleChange}
          isDisabled
          type={InputType.TEXT}
          label={'Real Name'}
          hint={'First name & Last name'}
        />
        <Input
          id={'email'}
          value={testUser.email}
          onChange={handleChange}
          type={InputType.EMAIL}
          isDisabled
          label={'Email'}
        />
      </div>
    </main>
  );
};

export default MyAccountPage;
