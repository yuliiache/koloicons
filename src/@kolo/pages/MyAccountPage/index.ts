import MyAccountPage from './MyAccountPage';

export default MyAccountPage;
