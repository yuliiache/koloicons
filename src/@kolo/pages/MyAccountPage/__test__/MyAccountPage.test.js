import React from 'react';
import {render, screen} from '@testing-library/react';

import '@testing-library/jest-dom';

import MyAccountPage from '../MyAccountPage';

const mockUser = {
  firstName: 'John',
  lastName: 'Doe',
  email: 'john.doe@mail.com',
};

describe('MyAccountPage Component', () => {
  it('renders correctly and matches snapshot', () => {
    const {asFragment} = render(<MyAccountPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the decorative circle with the first letter of the firstName', () => {
    render(<MyAccountPage />);
    const firstLetterElement = screen.getByText(mockUser.firstName.charAt(0));
    expect(firstLetterElement).toBeInTheDocument();
    expect(firstLetterElement).toHaveClass('circle');
  });

  it('displays the correct first name and email', () => {
    render(<MyAccountPage />);
    expect(screen.getByText(mockUser.firstName)).toBeInTheDocument();
    expect(screen.getByText(mockUser.email)).toBeInTheDocument();
  });

  it('renders input fields with correct values and properties', () => {
    render(<MyAccountPage />);
    const nameInput = screen.getByLabelText('Real Name');
    const emailInput = screen.getByLabelText('Email');

    expect(nameInput).toHaveValue(`${mockUser.firstName} ${mockUser.lastName}`);
    expect(emailInput).toHaveValue(mockUser.email);
    expect(nameInput).toBeDisabled();
    expect(emailInput).toBeDisabled();
  });
});
