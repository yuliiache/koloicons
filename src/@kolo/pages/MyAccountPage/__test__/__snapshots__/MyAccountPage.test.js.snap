// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyAccountPage Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="header"
  >
    <div
      class="headerContainer"
    >
      <div
        class="circle"
      >
        J
      </div>
      <section
        class="userInfo"
      >
        <h1>
          John
        </h1>
        <p>
          john.doe@mail.com
        </p>
      </section>
    </div>
  </section>
  <div
    class="inputGroup"
  >
    <div
      class="inputGroup"
    >
      <label
        class="inputGroupLabel"
        for="name"
      >
        Real Name
      </label>
      <span
        class="inputGroupHint"
      >
        First name & Last name
      </span>
      <input
        class="inputGroupInput"
        disabled=""
        id="name"
        type="text"
        value="John Doe"
      />
    </div>
    <div
      class="inputGroup"
    >
      <label
        class="inputGroupLabel"
        for="email"
      >
        Email
      </label>
      <input
        class="inputGroupInput"
        disabled=""
        id="email"
        type="email"
        value="john.doe@mail.com"
      />
    </div>
  </div>
</DocumentFragment>
`;
